db/collection.go

```diff
@@ -526,6 +529,283 @@ func validateUpdateSchemaFields(
+	decoder.DisallowUnknownFields()
+	err = decoder.Decode(&newColsByID)
+	if err != nil {
+		return err
+	}
+
+	err = db.validateCollectionChanges(existingColsByID, newColsByID)
+	if err != nil {
+		return err
+	}
+
+	for _, col := range newColsByID {
+		_, err := description.SaveCollection(ctx, txn, col)
+		if err != nil {
+			return err
+		}
+	}
+
+	return db.loadSchema(ctx, txn)
+}
+
+var patchCollectionValidators = []func(
+	map[uint32]client.CollectionDescription,
+	map[uint32]client.CollectionDescription,
+) error{
+	validateCollectionNameUnique,
+	validateSingleVersionActive,
+	validateSourcesNotModified,
+	validateIndexesNotModified,
+	validateFieldsNotModified,
+	validateIDNotZero,
+	validateIDUnique,
+	validateIDExists,
+	validateRootIDNotMutated,
+	validateSchemaVersionIDNotMutated,
+	validateCollectionNotRemoved,
+}
+
+func (db *db) validateCollectionChanges(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	for _, validators := range patchCollectionValidators {
+		err := validators(oldColsByID, newColsByID)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateCollectionNameUnique(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	names := map[string]struct{}{}
+	for _, col := range newColsByID {
+		if !col.Name.HasValue() {
+			continue
+		}
+
+		if _, ok := names[col.Name.Value()]; ok {
+			return NewErrCollectionAlreadyExists(col.Name.Value())
+		}
+		names[col.Name.Value()] = struct{}{}
+	}
+
+	return nil
+}
+
+func validateSingleVersionActive(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	rootsWithActiveCol := map[uint32]struct{}{}
+	for _, col := range newColsByID {
+		if !col.Name.HasValue() {
+			continue
+		}
+
+		if _, ok := rootsWithActiveCol[col.RootID]; ok {
+			return NewErrMultipleActiveCollectionVersions(col.Name.Value(), col.RootID)
+		}
+		rootsWithActiveCol[col.RootID] = struct{}{}
+	}
+
+	return nil
+}
+
+func validateSourcesNotModified(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	for _, newCol := range newColsByID {
+		oldCol, ok := oldColsByID[newCol.ID]
+		if !ok {
+			continue
+		}
+
+		// DeepEqual is temporary, as this validation is temporary, for example soon
+		// users will be able to be able to change the migration
+		if !reflect.DeepEqual(oldCol.Sources, newCol.Sources) {
+			return NewErrCollectionSourcesCannotBeMutated(newCol.ID)
+		}
+	}
+
+	return nil
+}
+
+func validateIndexesNotModified(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	for _, newCol := range newColsByID {
+		oldCol, ok := oldColsByID[newCol.ID]
+		if !ok {
+			continue
+		}
+
+		// DeepEqual is temporary, as this validation is temporary
```

Comment:
question: Why is this temporary? Will the use of reflection package be removed?

Reply:
The validation (function) is temporary, which of course means that if this function is removed, the call to DeepEqual will be removed with it.

Reply:
> The validation (function) is temporary, which means that if this function is removed, the call to DeepEqual will be removed with it.

Why is it temporary?

Reply:
Because the function is validating that indexes have not been modified, and we very much want them to be mutable in the future.

Reply:
Thanks that makes so much sense.

Please document this